### ISSUE TITLE
NEXT_LOCALE does not work with sub-path

### ISSUE THREAD

**Issue by elbopha** [vercel/next.js#21693]
**What version of Next.js are you using?**

10.0.6

**What version of Node.js are you using?**

14.15.0

**What browser are you using?**

Safari

**What operating system are you using?**

macOS

**How are you deploying your application?**

next start

**Describe the Bug**

Hello,

`NEXT_LOCALE` cookie is set to `en` for example, assuming `en` is not the default locale.

When I go to `/`, it will redirect to `/en`.

Now, when I go to `/about`, it didn't redirect to `/en/about`.
It does not work for any sub-path of `/`.

Did I forget something?

Regards,

**Expected Behavior**

`/` => `/en`, work fine
`/about` => `/en/about`, should work as well like `/`, for any sub-path

**To Reproduce**

- Set cookie `NEXT_LOCALE` to `en`
- Go to a sub-path of `/` like `/about`

**Comment by sidou01** [comment]
I'm having the same issue.

**Comment by sebastienlabine** [comment]
This is an expected behaviour as per the [documentation ](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie).

However, I really don't understand the reason behind it. If the user select a locale, it's for the whole app, not only for when he's going to the home page. This is pretty much a deal breaker for overriding the default locale.

@timer is there a specific reason behind this? Could we at least get an option to apply the cookie overwrite on all page?

**Comment by ronfogel** [comment]
I also agree that this logic should be applied for all the routes (except for api routes obviously) and would love to see it implemented.

**Comment by Amethystafyy282** [comment]
> This is an expected behaviour as per the [documentation ](https://nextjs.org/docs/advanced-features/i18n-routing#leveraging-the-next_locale-cookie).
> 
> However, I really don't understand the reason behind it. If the user select a locale, it's for the whole app, not only for when he's going to the home page. This is pretty much a deal breaker for overriding the default locale.
> 
> @Timer is there a specific reason behind this? Could we at least get an option to apply the cookie overwrite on all page?
> 
> This issue has been reported by several members. (#20771, #22863, #22375, #20375 )

I think i figured out why they did that.

1: For UX: User would choose language first and surfing the whole app.
2: If u use user's locale for the whole app, then custom locale put in url such as `localhost/en` is not work
3: Further from 2, u can`t custom render any different locale in any other specific page

A workaround that i've done is created a `withLocale` HOC just like this:

    const userLocale = req.cookies && req.cookies['NEXT_LOCALE'];

    return serverSideTranslations(userLocale || context.locale); //context.locale === defaultLocale in config.js if localeDetection === false

Pros: user's locale will be the priority, just like we asked for
Cons: 
+ its always render locale in cookie even though u intentionally put locale behind your host just like this `localhost/en`
+ you have to clean it when user's session is gone

that's all i've been going around for today :D
